[BUG] Forcing C++ struct to be a const when used as a temporary in methods that take const T&

### Describe the bug

Essentially, If we have some C++ struct `IntContainer`, and it's used as a temporary to a method that takes `const T&`, for example `vec.push_back(IntContainer(...))`, we'll initialize `IntContainer` as a `const` with:

```
struct IntContainer const __pyx_t_1;
```

But, we would then run into this C++ error:

```
vector_const.cpp:2631:29: error: default initialization of an object of const type 'const struct IntContainer' without a user-provided default constructor
 2631 |   struct IntContainer const __pyx_t_1;
      |                             ^
      |                                      {}
```

And we cannot assign to it later on:
```
vector_const.cpp:2645:20: error: cannot assign to variable '__pyx_t_1' with const-qualified type 'const struct IntContainer'
 2645 |   __pyx_t_1.number = __pyx_v_number;
      |   ~~~~~~~~~~~~~~~~ ^
```

This only occurs if we create `IntContainer` in-line, but it's fixed if we pull it into a variable:

```
i = IntContainer(...)
vec.push_back(i)
```

### Code to reproduce the behaviour:

https://github.com/pilleye/cython/commit/8ab76a06915465d925489d5719e6f036aced2092 


```cython
python runtests.py -3 --backends cpp -vv vector_const --no-cleanup
```


### Expected behaviour

Both cases should be the same, both should not error.

### OS

Mac

### Python version

3.12

### Cython version

`master`

### Additional context

_No response_

--- a/Cython/Compiler/ExprNodes.py
+++ b/Cython/Compiler/ExprNodes.py
@@ -9578,7 +9578,7 @@ class DictNode(ExprNode):
         return False
 
     def coerce_to(self, dst_type, env):
-        dst_type = PyrexTypes.remove_cv_ref(dst_type, True)
+        dst_type = PyrexTypes.remove_cv_ref(dst_type, remove_fakeref=True)
         if dst_type.is_pyobject:
             self.release_errors()
             if self.type.is_struct_or_union:

--- a/Cython/Compiler/ExprNodes.py
+++ b/Cython/Compiler/ExprNodes.py
@@ -6044,6 +6044,9 @@ class CallNode(ExprNode):
             self.type = py_object_type
 
     def analyse_as_type_constructor(self, env):
+        """
+        Returns a replacement node or None
+        """
         type = self.function.analyse_as_type(env)
         if type and type.is_struct_or_union:
             args, kwds = self.explicit_args_kwds()
@@ -6052,11 +6055,10 @@ class CallNode(ExprNode):
                 items.append(DictItemNode(arg.pos, key=UnicodeNode(arg.pos, value=member.name), value=arg))
             if kwds:
                 items += kwds.key_value_pairs
-            self.key_value_pairs = items
-            self.__class__ = DictNode
-            self.analyse_types(env)    # FIXME
-            self.coerce_to(type, env)
-            return True
+
+            node = DictNode(self.pos, key_value_pairs=items)
+            node = node.analyse_types(env).coerce_to(type, env)
+            return node
         elif type and type.is_cpp_class:
             self.args = [ arg.analyse_types(env) for arg in self.args ]
             constructor = type.scope.lookup("<init>")
@@ -6069,7 +6071,7 @@ class CallNode(ExprNode):
             self.function.set_cname(type.empty_declaration_code())
             self.analyse_c_function_call(env)
             self.type = type
-            return True
+            return self
 
     def function_type(self):
         # Return the type of the function being called, coercing a function
@@ -6172,8 +6174,8 @@ class SimpleCallNode(CallNode):
         if self.analysed:
             return self
         self.analysed = True
-        if self.analyse_as_type_constructor(env):
-            return self
+        if (as_type_constructor := self.analyse_as_type_constructor(env)) is not None:
+            return as_type_constructor
         self.function.is_called = 1
         self.function = self.function.analyse_types(env)
         function = self.function
@@ -7170,8 +7172,8 @@ class GeneralCallNode(CallNode):
         return self.positional_args.args, self.keyword_args
 
     def analyse_types(self, env):
-        if self.analyse_as_type_constructor(env):
-            return self
+        if (as_type_constructor := self.analyse_as_type_constructor(env)) is not None:
+            return as_type_constructor
         self.function = self.function.analyse_types(env)
         if not self.function.type.is_pyobject:
             if self.function.type.is_error:
@@ -9576,6 +9578,7 @@ class DictNode(ExprNode):
         return False
 
     def coerce_to(self, dst_type, env):
+        dst_type = PyrexTypes.remove_cv_ref(dst_type, True)
         if dst_type.is_pyobject:
             self.release_errors()
             if self.type.is_struct_or_union:

--- a/tests/run/cpp_struct_refs.pyx
+++ b/tests/run/cpp_struct_refs.pyx
@@ -1,0 +1,18 @@
+# mode: run
+# tag: cpp
+
+from libcpp.vector cimport vector
+
+cdef struct S:
+    int number
+
+def construct_and_push_back(int number):
+    """
+    >>> construct_and_push_back(5)
+    5
+    """
+    v = vector[S]()
+    # The struct should be constructed with a type of S,
+    # rather than a type of const S&
+    v.push_back(S(number))
+    return v.back().number

